agent/agent.go

```diff
@@ -3465,4 +3653,141 @@ func (a *Agent) registerCache() {
+	return a.State
+}
+
+// rerouteExposedChecks will inject proxy address into check targets
+// Future calls to check() will dial the proxy listener
+// The agent stateLock MUST be held for this to be called
+func (a *Agent) rerouteExposedChecks(serviceID string, proxyAddr string) error {
+	for _, c := range a.checkHTTPs {
+		if c.ServiceID != serviceID {
+			continue
+		}
+		port, err := a.listenerPortLocked(serviceID, string(c.CheckID))
+		if err != nil {
+			return err
+		}
+		c.ProxyHTTP = httpInjectAddr(c.HTTP, proxyAddr, port)
+	}
+	for _, c := range a.checkGRPCs {
+		if c.ServiceID != serviceID {
+			continue
+		}
+		port, err := a.listenerPortLocked(serviceID, string(c.CheckID))
+		if err != nil {
+			return err
+		}
+		c.ProxyGRPC = grpcInjectAddr(c.GRPC, proxyAddr, port)
+	}
+	return nil
+}
+
+// resetExposedChecks will set Proxy addr in HTTP checks to empty string
+// Future calls to check() will use the original target c.HTTP or c.GRPC
+// The agent stateLock MUST be held for this to be called
+func (a *Agent) resetExposedChecks(serviceID string) {
+	ids := make([]string, 0)
+	for _, c := range a.checkHTTPs {
+		if c.ServiceID == serviceID {
+			c.ProxyHTTP = ""
+			ids = append(ids, string(c.CheckID))
+		}
+	}
+	for _, c := range a.checkGRPCs {
+		if c.ServiceID == serviceID {
+			c.ProxyGRPC = ""
+			ids = append(ids, string(c.CheckID))
+		}
+	}
+	for _, checkID := range ids {
+		delete(a.exposedPorts, listenerPortKey(serviceID, checkID))
+	}
+}
+
+// listenerPort allocates a port from the configured range
+// The agent stateLock MUST be held when this is called
+func (a *Agent) listenerPortLocked(svcID, checkID string) (int, error) {
+	key := listenerPortKey(svcID, checkID)
+	if a.exposedPorts == nil {
+		a.exposedPorts = make(map[string]int)
+	}
+	if p, ok := a.exposedPorts[key]; ok {
+		return p, nil
+	}
+
+	allocated := make(map[int]bool)
+	for _, v := range a.exposedPorts {
+		allocated[v] = true
+	}
+
+	var port int
+	for i := 0; i < a.config.ExposeMaxPort-a.config.ExposeMinPort; i++ {
+		port = a.config.ExposeMinPort + i
+		if !allocated[port] {
+			a.exposedPorts[key] = port
+			break
+		}
+	}
+	if port == 0 {
+		return 0, fmt.Errorf("no ports available to expose '%s'", checkID)
+	}
+
+	return port, nil
+}
+
+func listenerPortKey(svcID, checkID string) string {
+	return fmt.Sprintf("%s:%s", svcID, checkID)
+}
+
+// grpcInjectAddr injects an ip and port into an address of the form: ip:port[/service]
+func grpcInjectAddr(existing string, ip string, port int) string {
+	r := regexp.MustCompile("(.*)((?::)(?:[0-9]+))(.*)$")
+
+	portRepl := fmt.Sprintf("${1}:%d${3}", port)
+	out := r.ReplaceAllString(existing, portRepl)
+
+	addrRepl := fmt.Sprintf("%s${2}${3}", ip)
+	out = r.ReplaceAllString(out, addrRepl)
+
+	return out
+}
+
+// httpInjectAddr injects a port then an IP into a URL
+func httpInjectAddr(url string, ip string, port int) string {
+	r := regexp.MustCompile(`^(http[s]?://)(\[.*?\]|\[?[\w\-\.]+)(:\d+)?([^?]*)(\?.*)?$`)
```

Comment:
As this is constant, you could move this up to be a package local variable like `httpInjectAddrPattern`

Reply:
done